Gitlab Integration

gokmen opened the pull request:
![](http://g.recordit.co/BDMbhwgxPD.gif)

Full screencasts: 
- http://recordit.co/BDMbhwgxPD
- http://recordit.co/By0qiz1ClC

```
λ ~/Code/koding/ gitlab ./run nodetestfiles servers/lib/server/handlers/api/gitlab


  server.handlers.api.gitlab
    ✓ should handle project_create requests
    ✓ should handle project_destroy requests
    ✓ should handle project_rename requests
    ✓ should handle project_transfer requests
    ✓ should handle key_create requests
    ✓ should handle key_destroy requests
    ✓ should handle group_create requests
    ✓ should handle group_destroy requests
    ✓ should handle tag_push requests
    ✓ should handle push requests

  server.handlers.api.gitlab.user
    server.handlers.api.gitlab.user.create
      ✓ should create a user (769ms)
    server.handlers.api.gitlab.user.destroy
      ✓ should destroy a user
    server.handlers.api.gitlab.user.add_to_team
      ✓ should add user to given team
    server.handlers.api.gitlab.user.remove_from_team
      ✓ should remove user from given team
    server.handlers.api.gitlab.user.add_to_group
      ✓ should add user to given group
    server.handlers.api.gitlab.user.remove_from_group
      ✓ should remove user from given group


  16 passing (2s)
```

On the GitLab-CE side we do have another MR waiting for https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/4769

gokmen committed:
Account: re-enable Externals for now

gokmen committed:
Landing: enable login with GitLab

gokmen committed:
Config: define GitLab configs for integration

gokmen committed:
Config: main configs: allow to overwrite hostname from environment

gokmen committed:
GitLab: oauth implementation initial commit

gokmen committed:
OAuth: enable GitLab OAuth flow

gokmen committed:
GitLab: Handler: initial commit for the handler

gokmen committed:
GitLab: Handler: for Group events

gokmen committed:
GitLab: Handler: for User events

gokmen committed:
GitLab: Handler: for Tag events

gokmen committed:
GitLab: Handler: for Push events

gokmen committed:
GitLab: Handler: for Key events

gokmen committed:
GitLab: Handler: for Project events

gokmen committed:
GitLab: GenericHandler: base class for all handlers

gokmen committed:
GitLab: Event data for reference

gokmen committed:
GitLab: Require all event handlers

gokmen committed:
Server: handle -/api/gitlab events on gitlab handler

gokmen committed:
GitLab: Fix linter issues

gokmen committed:
GitLab: update push/tag data parsers to use try-catch

gokmen committed:
GitLab: events export events

gokmen committed:
GitLab: separate utils into it's own helper

gokmen committed:
GitLab: sample test data for tests

gokmen committed:
GitLab: test helpers for unit tests

gokmen committed:
GitLab: handle requests correctly and use proper method call

gokmen committed:
GitLab: tests for main handlers, just tests handle statuses for now

gokmen committed:
OAuth: disable cyclomatic complexity for now

fixme @gokmen

gokmen committed:
Credentials.default: add systemHookToken

gokmen committed:
GitLab: add more sample test data for tests

gokmen committed:
JSession: update foreignAuth schema to add GitLab

gokmen committed:
OAuth: put fixed protocol for now, this going to be changed

gokmen committed:
Api: handler: move handleUsername helper into helpers

gokmen committed:
Api: handler: move validateUsername helper into helpers

gokmen committed:
GitLab: update sample test data

gokmen committed:
GitLab: Main handler log handler errors

gokmen committed:
GitLab: update genericHandler to have Bongo and data validators

gokmen committed:
GitLab: User: create handler for user_create events

gokmen committed:
GitLab: User: destroy handler for user_destroy events

gokmen committed:
GitLab: Test helpers for generic requests on tests

gokmen committed:
GitLab: User: Tests for create and destroy events

gokmen committed:
GitLab: User: Tests for not implemented handlers as they are not implemented

gokmen committed:
GitLab: Use token from config

gokmen committed:
GitLab: system hooks tied to config status

gokmen committed:
Run: remove create default workspace command

gokmen committed:
Cosmetic: generic tidy up

gokmen committed:
JSession: createSession: allow creating session for a given group

gokmen committed:
JSession: fetchSession: add support for group context

all related calls also updated for new api

gokmen committed:
SetSession: set the clientId on cookie again for following calls

Before this one we were creating n different sessions based on the
following requests. Once user hits the server without a clientId we
were creating the session in here but not updating the cookie for the
rest of the code to have clientId information, so, on each step
which requires session, it was re-creating again

gokmen committed:
OAuth: use the port for oauth endpoint if provided

gokmen committed:
SetSession: set session group information based on the req.subdomains

gokmen committed:
Client: Proxifier: use koding.com proxy services for on-premise installs

gokmen committed:
Config: use provided host argument for domains.main

gokmen committed:
Landing: utils: use domains.main for team name parser from url

gokmen committed:
Landing: update greens site-wide

gokmen committed:
Landing: Makefile: add dev target for default landing page build

gokmen committed:
TeamLoginTab: use same style for both sign in and gitlab link

gokmen committed:
LoginForm: update style for gitlab login

gokmen committed:
Landing: Makefile: use build only for default target

gokmen committed:
LoginForm: Style update completed with icons

gokmen committed:
LoginForm: Style updates based on design team feedback

gokmen committed:
Default config: add default team slug for gitlab integration

gokmen committed:
Webserver: provide gitlab runtimeConfig on payload

gokmen committed:
Landing: Login: remove gitlab related views from solo login

gokmen committed:
LoginForm: all gitlab login views under gitlabLogin view

gokmen committed:
Landing: if landed group is same with gitlab group export gitlab options

gokmen committed:
TeamLoginTab: show gitlabLogin if the team has gitlab integration

gokmen committed:
Routes: handleOAuth: open team login for ouath handler endpoint

with a graceful notification

gokmen committed:
GenerateRuntimeConfig: use credentials.gitlab if provided

gokmen committed:
GitLabProvider: use upstream gitlab package

gokmen committed:
GitProvider: Constants: use .koding.yml and .koding.README instead

gokmen committed:
StackEditor: use existing config content from stack template if possible

gokmen committed:
GitLabProvider: getPrivateToken to get private token with oauth token

gokmen committed:
GitLabProvider: if oauth token exists then try to fetch from that one

gokmen committed:
GitLabProvider: parseRepo helper to extract repo url info from given data

gokmen committed:
GitLabProvider: try to parseRepo if url is not given, ignore other urls

gokmen committed:
GitLabProvider: change provided data as template and readme objects

gokmen committed:
GithubProvider: update provided data as objects

gokmen committed:
GitProvider.test: update provided data as objects in tests as well

gokmen committed:
GitLabProvider: make sure provided branch exists on remote

gokmen committed:
GitProvider: instead of importing based on the url, take action based on the requested provider

gokmen committed:
GitProvider: on stack create keep repo related information in config

gokmen committed:
Sidebar: show open on GitLab link for stacks created from a repository

gokmen committed:
Sidebar: show open on GitLab link for stack templates created from a repo

gokmen committed:
GitProvider: Constants: use README.md instead

gokmen committed:
Home: RouteHandler: handle Home/Stacks/run/OWNER/REPO urls

gokmen committed:
GitLabProvider: handle private token fetch, no need for second instance

gokmen committed:
GitProvider: if stack template exists with given remote data, return it

gokmen committed:
RouteHandler: defer the operation

gokmen committed:
GitProvider: Test: generate new url for each test

gokmen committed:
Home: Account: Add Integrations for GitLab

gokmen committed:
GitLabProvider: keep fetched privateToken in foreignAuth and re-use it

gokmen committed:
Home: Account: Integrations: fix linter issues

gokmen committed:
GitProvider.test: re-generate new repo owner on each tests

gokmen committed:
RouteHandler: handle branch details as well

gokmen committed:
GitProvider: Use branch information on remoteDetails for querying

gokmen commented:
ref: #8220

gokmen committed:
Landing: utils: replace all dots in baseDomain

gokmen committed:
Config: get all GitLab configuration from environment if provided

gokmen committed:
SetSession: handle xip.io domains correctly if it's used as hostname

gokmen committed:
JGroupData: for private group meta data

gokmen committed:
JGroup: fetchData{,At} helpers for handling JGroupData over JGroup inst.

gokmen committed:
JGroup: fix linter issues

cihangir commented on servers/lib/server/gitlab_callback.coffee:
```diff
@@ -0,0 +1,97 @@
+http       = require 'http'
```
i would probably use `request` package

cihangir commented on servers/lib/server/handlers/api/gitlab/generichandler.coffee:
```diff
@@ -0,0 +1,41 @@
+
+      else
+
+        { error: apiErrors.internalError }
```
this case does not seem like an internal error

cihangir commented on servers/lib/server/handlers/api/gitlab/user.coffee:
```diff
@@ -0,0 +1,150 @@
+
+      (next) ->
+        context = { group: 'gitlab' }
+        bongo.fetchClient 1, context, (client_) ->
```
having 1 here is problematic. Could you pass clientId from `req` into this func? ( better to do smth like generateFakeClientFromReq does)

gokmen replied:
I don't want to include `req` into this level, tried to create an abstract api which only does the operation itself, not relies on anything. I can make `fetchClient` to act based on parameters or I can make the `1` as `false` and handle the flow in that way. But I don't want to carry `req` up to here, think this as a 1st level of api for user creation, I'll move this out to the first level api implementation package (once we've something like that)

cihangir commented on servers/lib/server/handlers/api/gitlab/user.coffee:
```diff
@@ -0,0 +1,150 @@
+      (next) ->
+        JSession.createNewSession { username, groupName }, (err, session) ->
+          if err or not session
+            return next err ? { message: 'user not exists' }
```
we do not validate if user exists on session creation

cihangir commented on servers/lib/server/setsession.coffee:
```diff
@@ -1,26 +1,37 @@
     # update clientId cookie
     updateCookie req, res, result.session
 
+    req.cookies.clientId = result.session.clientId
```
what can i do sometimes :/

👍

cihangir commented on workers/social/lib/social/models/gitprovider/constants.coffee:
```diff
@@ -2,5 +2,5 @@ module.exports =
   GITLAB_HOST     : 'gitlab.com'
-  TEMPLATE_PATH   : 'Koding.yaml'
-  README_PATH     : 'Koding.README'
+  TEMPLATE_PATH   : '.koding.yml'
```
can we also support `yaml` sometime in the future 🙏 

( i hate when CI services fail to accept yaml )

gokmen replied:
right, we can add multiple template file supports which I didn't want to do that right now since it causes n times more api calls for getting the file content. but it's doable of course.

cihangir commented on workers/social/lib/social/models/gitprovider/gitlabprovider.coffee:
```diff
@@ -1,72 +1,167 @@
   parseImportUrl: (url) ->
 
+    # if user/repo/branch provided as url we will use
+    if not (/^http(|s)\:\/\//.test url) and (url.split '/').length in [2, 3]
```
😕

gokmen replied:
no worries :) I'll get all details as query strings instead, in this way it's not extendable.

cihangir commented on workers/social/lib/social/models/oauth.coffee:
```diff
@@ -1,3 +1,5 @@
+# coffeelint: disable=cyclomatic_complexity
```
😠

cihangir commented:
mostly LGTM, we can add/implement/fix things on the road


fyi: I did not check the files under client/*

szkl commented on client/app/lib/util/proxifier.coffee:
```diff
@@ -35,7 +35,11 @@ module.exports = class Proxifier
-    baseURL = "#{protocol}//#{subdomain}.#{globals.config.domains.base}/-"
+    # Since proxy is a service from koding.com we need to use it
+    # for on-premise version as well. otherwise all the on-premises
+    # users needs to run proxy server on their own ~GG
```
I am going to work on running user proxy in default configuration next.

szkl commented on client/app/lib/util/routeHandler.coffee:
```diff
@@ -29,6 +33,51 @@ module.exports = (options) ->
+    { section, action, identifier, extra, branch } = params
+
+    repo = "#{identifier}/#{extra}"
+    repo = "#{repo}/#{branch}"  if branch
```
```coffeescript
repo += "/#{branch}"  if branch
```

gokmen replied:
we don't like concatenation here https://github.com/koding/styleguide-coffeescript#use-string-interpolations-instead-of-string-concatenation :)

gokmen committed:
SetSession: use hostname instead host will be deprecated soon

gokmen committed:
GitLab: Oauth: use request instead of http package

gokmen committed:
GenericHandler: use notImplementedError instead code: 501

gokmen committed:
GitLab: Utils: make sure to keep string copy of token

gokmen committed:
GitLab: User: updates based on reviews

gokmen committed:
Models: OAuth: fix cyclomatic complexity, tidy up code

gokmen committed:
GenerateKonfig: enable hooks correctly

gokmen committed:
GenerateKonfig: fix credential variable

gokmen committed:
RealtimeController: disable on default environment

gokmen committed:
Credentials: Default: update postgresql credentials

gokmen committed:
Config: GenerateKonfig: check for value instead of existence

gokmen committed:
Config: Default: update customDomain.public

gokmen committed:
Config: Default: disable algoliaconnector, paymentwebhook, gatekeeper, vmwatcher workers

gokmen committed:
GitLab: User event handler: use the team from KONFIG

gokmen committed:
README.GitLab: Initial commit

gokmen committed:
HomeAppView: fix styling on GitLab Integration section

gokmen committed:
Fix linter issue, move README.gitlab into root

gokmen committed:
README: readme updates, thanks to @sinan

gokmen committed:
GenerateKonfig: fix gitlab tests for dev environments

gokmen committed:
Config: add gitlab into optional features list, enabled for default env.

gokmen committed:
Utils: isFeatureEnabled: add additionalcheck support for custom features

Features will be disabled based on config and some other conditional
checks which we can simply defined them here.

We disabled gitlab feature for other teams but the one in the config.

gokmen committed:
Landing: disable single-sign-on to gitlab feature on production

gokmen committed:
Landing: remove farewell-solo route on Login request which is default

This creates lots of issues currently since Login is the default route
on most of the places and it ends up with the blog post. Since it's
already removed as a feature on production we don't need to worry about
this redirection, instead we can guide these users who has an issue with
solo logins in support channels.

gokmen committed:
GitLab: OAuth: provide more information about the oauth process on error

gokmen committed:
Home: handle OAuth redirection on Home app instead of deprecated Account

gokmen committed:
Home, GitLab OAuth: update logs, use direct link for now

gokmen committed:
Home: pass query string to paneView if there is an handler exists for.

gokmen committed:
Home: Import: stack importer initial commit with all required features

gokmen committed:
Utils: objectToString: htmlDecode everything and replace new lines

gokmen committed:
OutputView: no need for JView feature-set

gokmen committed:
OutputView: set of improvements, delegate scroll, optional separator

gokmen committed:
HomeStackImport: update scroll on content change, style updates

gokmen committed:
HomeStackImport: show message on empty state for import pane

gokmen committed:
HomeStackImport: show an action button based on the import status

gokmen committed:
ObjectToString: fix linter issue

gokmen committed:
GitProvider: fetchConfig to fetch given provider's provided config

gokmen committed:
Github: getConfig: simply active host for Github

gokmen committed:
GitLab: getConfig: return config and host of active gitlab instance

This will only return valid config if request is being made from active
gitlab team.

gokmen committed:
Landing: handle oauth links correctly

gokmen committed:
HomeStackImport: use active gitlab link from provided config

gokmen commented:
This is ready to go waiting reviews from @koding/developers ! \o/

usirin commented on client/landing/Makefile:
```diff
@@ -1,3 +1,9 @@
+all: build
+
+dev:
+	@gulp --cwd site.landing --exportDir=./../../../website/a/site.landing/ --devMode
```
Thank you <3

sinan commented on client/app/lib/components/sidebar/index.coffee:
```diff
@@ -95,7 +95,9 @@ module.exports = class Sidebar extends React.Component
-
+      when 'Open on GitLab'
+        remoteUrl = draft.getIn ['config', 'remoteDetails', 'originalUrl']
+        window.open remoteUrl, '_blank'
```
you can keep a reference here to the window opened for later operations.

sinan commented on client/app/lib/util/isFeatureEnabled.coffee:
```diff
@@ -1,14 +1,21 @@
 module.exports = (feature) ->
 
   isDisabled = globals.config.disabledFeatures[feature]
+  additionalCheckPassed = additionalChecks[feature]?() ? true
```
minor: `additionalChecksPassed`

gokmen replied:
I kept it singular on purpose since it's supporting only one additional check function at the moment, once I add multiple checks support I was planning to update it like that. But we can do it before hand :)

sinan commented on client/home/lib/stacks/homestacksimport.coffee:
```diff
@@ -0,0 +1,151 @@
+
+    @addSubView new kd.CustomHTMLView
+      tagName: 'h2'
+      partial: "Stack Importer"
```
minor: lint error `"` > `'`

sinan replied:
but it seems to have passed the tests maybe we don't check for the quotes anymore? cc/ @usirin @hakankaradis

usirin replied:
The files under `home/stacks` are being [ignored by coffeelint](https://github.com/koding/koding/blob/master/.coffeelintignore#L50).

sinan commented on client/home/lib/stacks/homestacksimport.coffee:
```diff
@@ -0,0 +1,151 @@
+        This tool is intended to use with GitLab integration which is enabled
+        for this team on <a href="#{_host}">#{host}</a>. You can use action
+        endpoints provided in your GitLab service to be able to use this tool
+        to try your projects on Koding.
```
```coffee
"""
This tool is intended to be used with GitLab. The integration is enabled
for this team at <a href="#{_host}">#{host}</a>. To be able to run your
projects on Koding, please use action endpoints provided in your GitLab service.
"""
```

sinan commented on config/generateRuntimeConfig.coffee:
```diff
@@ -29,11 +29,13 @@ module.exports = (KONFIG, credentials, options) ->
+    troubleshoot         : { idleTime: 1000 * 60 * 60, externalUrl: 'https://s3.amazonaws.com/koding-ping/healthcheck.json' }
     stripe               : { token: credentials.stripe.publicToken }
-    broker               : { uri: "/subscribe" }
-    google               : { apiKey: 'AIzaSyDiLjJIdZcXvSnIwTGIg0kZ8qGO3QyNnpo' }
```
recycled?

gokmen replied:
I hope :) https://github.com/koding/koding/blob/master/config/generateRuntimeConfig.coffee#L36

sinan commented:
Looks good to me too! 💯

gokmen commented on client/features/lib/styl/features.styl:
```diff
@@ -119,7 +119,7 @@ body.features
           font-size           21px
           line-height         22px
-          color               #27ae60
+          color               #48BA7D
```
Leaving it as is since all the greens updated, instead of adding one more commit.

gokmen committed:
HomeStackImport: copy updates and some review issues fixed

gokmen committed:
GitProvider: update parseImportUrl to parseImportData to count repo as well

gokmen committed:
GitProvider: update corresponding tests for parseImportData changes

gokmen committed:
isFeatureEnabled: /additionalCheckPassed/additionalChecksPassed/i

gokmen committed:
LinkController: openOrFocus to provide referenced open feature for links

gokmen committed:
HomeStackImport: use linkController.openOrFocus where needed

gokmen committed:
ComputeController: helpers to find stacks/machines from remote data

We'll use these helpers to find related stacks/machines from provided
remote repository data such as commitId, branch etc.

gokmen committed:
GitLabProvider: injectQueryStrings: replace provided branch

This will allow us to use same stack template for different branches, so
user will provide the branch name with queryString while importing the
stack template from a repository.

gokmen committed:
HomeStackImport: open stack editor automatically

gokmen committed:
HomeStackImport: check for existing vms before processing repository

We'll check for existing vms to have same commitId then we will redirect
user to this machine instead of the same process again.

gokmen commented:
I've fixed all the pointed out issues in the PR and made some changes based on the GitLab requirements for better user experience, latest screencast can be found here http://recordit.co/By0qiz1ClC 

New features are:
 - Launch editor automatically once import completed
 - Launch existing vm if second request comes for the same repo/commitId
 - Support branch templating so we can use same stack template for the different branches of same repository

Related changes on GitLab side can be found here: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/4769#note_13880131

szkl commented:
👍

gokmen committed:
Style fixes for Safari

gokmen committed:
README.gitlab: Update readme

gokmen committed:
GitLabProvider: replace repo from queryString as well
